### PR TITLE
[Android 13+] Restore support of custom notification actions

### PR DESCRIPTION
**Pull request by Stypox**
#### What is it?
- [x] Bugfix (user facing)
- [ ] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
- Supersedes https://github.com/TeamNewPipe/NewPipe/pull/10580 and https://github.com/TeamNewPipe/NewPipe/pull/10567 because the **hacky workaround** to allow customizing 4 actions [turned out to cause problems](https://github.com/TeamNewPipe/NewPipe/pull/10580#discussion_r1435593734)
- Set notification actions only via `MediaSessionConnector` on Android 13+, and only via the notification builder on Android 12-, to save battery
- Notification actions in `MediaSessionPlayerUi` are only updated when something really changes, to avoid `setCustomActionProviders()` calling `invalidateMediaSessionPlaybackState()` more times as needed, which I figure might be costly
- Adapts the player notification settings page to allow customizing only two actions on Android 13+. The preference keys for the first three actions are actually still set, to avoid complicating the logic too much, and to allow seamless transitioning between Android 12- and 13+ back and forth (e.g. when a user exports and imports a database).
- Allows each notification slot to contain any possible action (which required changing how compact notification slot indices are computed, since the assumption that "Nothing" slots are always last does not hold anymore)
- Allows play/pausing from the notification when the player is buffering, which is in line with what happens in the main player ui
- Cleans up the code by extracting `NotificationSlot` from `NotificationActionsPreference`, and `NotificationActionData` from `NotificationUtil`

I tested a bit and caught some small inconsistencies in the code that now are fixed, and everything seems to work as expected.

#### Before/After Screenshots/Screen Record

<table>
  <tr>
    <th>Before notification</th>
    <th>After notification</th>
    <th>Settings in 12-</th>
    <th>Settings in 13+</th>
  </tr>
  <tr>
    <td>Before</td>
    <td>After</td>
    <td>After</td>
    <td>After</td>
  </tr>
</table>

#### Fixes the following issue(s)

- Fixes https://github.com/TeamNewPipe/NewPipe/issues/9764

#### APK testing

The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR. You can find more info and a video demonstration [on this wiki page](https://github.com/TeamNewPipe/NewPipe/wiki/Download-APK-for-PR).

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

File: app/src/main/java/org/schabi/newpipe/settings/custom/NotificationSlot.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        ((TextView) view.findViewById(R.id.notificationActionTitle))
+                .setText(SLOT_TITLES[i]);
+        view.findViewById(R.id.notificationActionClickableArea).setOnClickListener(
+                v -> openActionChooserDialog());
+    }
+
+    void setupCheckbox(final View view, final boolean isCompactSlotChecked) {
+        final CheckBox compactSlotCheckBox = view.findViewById(R.id.notificationActionCheckBox);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            // there are no compact slots to customize on Android 33+
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/settings/custom/NotificationSlot.java:92]: Either Android 13+ or API 33+
```suggestion
            // there are no compact slots to customize on Android 13+
```

**Comment by Stypox** [comment]
Is https://github.com/TeamNewPipe/NewPipe/pull/10580#issuecomment-1868289126 a bug of this PR? I think it's more of a consequence of AudricV's PR that fixed ANR popups by showing a notification asap. I know it's not perfect but it's better to show the notification asap to avoid getting errors for taking too much time to show the foreground service notification.

About cutting down actions: you're right but maybe some users would find some use for some of those options (even if they (partially) overlap with the already shown ones), so I wouldn't remove them as of now.

**Review by TobiGr — APPROVED** [review]
Code looks good to me. I did some testing and was not able to find any issues on API 33, 32, 29, 26

**Comment by dvalter** [comment]
For me it brings a huge regression compared to #10580 and pre-33 notifications.
On my devices (A13, LineageOS) it removes rewind/fast-forward from the default notification actions (and an option to configure force rewind/fast-forward for playlists).

It would be better to keep prev/next buttons semi-configurable for 33+ in settings rather than crippling their functionality.

**Comment by Stypox** [comment]
> For me it brings a huge regression compared to https://github.com/TeamNewPipe/NewPipe/pull/10580 and pre-33 notifications.

Yeah, I know... But that's how Google intended it, and I agree they made a bad choice there...

Unfortunately we can't add a configurable switch to enable the hacky workaround for #10580 for various reasons
- maintenance is difficult when there are too many settings, due to more unexpected bugs popping up
- most users wouldn't stumble upon the setting anyway, even if they need it, considering how niche this change is
- users that enable the setting and then forget about it will start arguing that their bluetooth controls don't work

**Comment by Stypox** [comment]
I guess that would be simple to do, yeah.

**Comment by dvalter** [comment]
> Yeah, I know... But that's how Google intended it, and I agree they made a bad choice there...

They rarely make something new without screwing up something good and old.

> users that enable the setting and then forget about it will start arguing that their bluetooth controls don't work

That made me think, I have an idea, how to improve the UX in the most broken case. I'll open a new issue then.

**Comment by dvalter** [comment]
@Stypox looks like #7880 is what I was thinking about.

Reassigning next/prev actions for the whole player should also do the trick with A13 notifications. And that may even be used in a "smart next/forward" mode to make notification less useless for a single video playback (which is 99% of my usage of NewPipe).

**Comment by Stypox** [comment]
Thanks for looking it up! Yeah #7880 makes more sense than an option just for the notification

**Comment by Stypox** [comment]
Ok

**Comment by einzel** [comment]
Where should I post if I am still encountering a problem to this? I am on Android 14 and my notification icon shows no control buttons and only forwards to the app or closes by swiping.

**Pull request merged**